File: Anul 2/Medii de Programare si Proiectare/Lab/mpp-proiect-csharp-Mothyque/ProiectMPP/ProiectMPP.Server/Service/TicketService.cs
using ProiectMPP.Common.Domain;
using ProiectMPP.Server.Repository;
using System;
using System.Collections.Generic;

namespace ProiectMPP.Server.Service
{
    public class TicketService : Service<int, Ticket>
    {
        private readonly ITicketRepository _ticketRepository;
        private readonly IMatchRepository _matchRepository;

        public TicketService(ITicketRepository ticketRepository, IMatchRepository matchRepository)
            : base(ticketRepository)
        {
            _ticketRepository = ticketRepository;
            _matchRepository = matchRepository;
        }

        public Ticket GetTicket(int clientId, int matchId)
        {
            int ticketId = _ticketRepository.GetTicketId(clientId, matchId);
            if (ticketId != -1)
            {
                return _ticketRepository.FindOne(ticketId);
            }
            return null;
        }

        public List<ClientTicketDTO> GetTicketByNameAndAddress(string clientName, string clientAddress)
        {
            return (List<ClientTicketDTO>)_ticketRepository.GetClientTicketsByCriteria(clientName, clientAddress);
        }

        public void RemoveTickets(int clientId, int matchId, int number)
        {
            for (int i = 0; i < number; i++)
            {
                int ticketId = _ticketRepository.GetTicketId(clientId, matchId);
                if (ticketId != -1)
                {
                    _ticketRepository.Delete(ticketId);
                }
            }
            UpdateObservers(); 
        }

        public void UpdateClientTickets(Client client, Match match, int diff)
        {
            if (diff > 0)
            {
                int currentAvailable = match.AvailableSeats;
                if (currentAvailable < diff)
                {
                    throw new ArgumentException("Not enough available seats for the match.");
                }

                for (int i = 0; i < diff; i++)
                {
                    int seatNumber = match.TotalSeats - currentAvailable + i + 1;
                    _ticketRepository.Save(new Ticket(client.Id, match.Id, "Seat " + seatNumber));
                }
            }
            else
            {
                RemoveTickets(client.Id, match.Id, Math.Abs(diff));
            }

            match.AvailableSeats -= diff;
            _matchRepository.Update(match);
            UpdateObservers();
        }

        public List<ClientTicketDTO> GetFilteredTickets(string name, string address)
        {
            return (List<ClientTicketDTO>)_ticketRepository.GetClientTicketsByCriteria(name, address);
        }
    }
}

File: Anul 2/Medii de Programare si Proiectare/Lab/mpp-proiect-java-Mothyque/backend/Server/src/main/java/org/example/service/TicketService.java
package org.example.service;

import org.example.domain.Client;
import org.example.domain.ClientTicketDTO;
import org.example.domain.Match;
import org.example.domain.Ticket;
import org.example.repository.MatchRepository;
import org.example.repository.TicketRepository;

import java.util.List;

public class TicketService extends Service<Integer, Ticket>
{
    private final TicketRepository ticketRepository;
    private final MatchRepository matchRepository;

    public TicketService(TicketRepository ticketRepository, MatchRepository matchRepository)
    {
        super(ticketRepository);
        this.ticketRepository = ticketRepository;
        this.matchRepository = matchRepository;
    }

    public Ticket getTicket(Integer clientId, Integer matchId)
    {
        int ticketId = ticketRepository.getTicketId(clientId, matchId);
        if (ticketId != -1)
        {
            return ticketRepository.findOne(ticketId).orElse(null);
        }
        return null;
    }

    public List<ClientTicketDTO> getTicketByNameAndAddress(String clientName, String clientAddress)
    {
        return  ticketRepository.getClientTicketsByCriteria(clientName, clientAddress);
    }

    public void removeTickets(Integer clientId, Integer matchId, int number)
    {
        for (int i = 0; i < number; i++)
        {
            int ticketId = ticketRepository.getTicketId(clientId, matchId);
            if (ticketId != -1)
            {
                ticketRepository.delete(ticketId);
            }
        }
        updateObservers();
    }

    public void updateClientTickets(Client client, Match match, int diff)
    {
        if (diff > 0)
        {
            int currentAvailable = match.getAvailableSeats();
            if (currentAvailable < diff)
            {
                throw new IllegalArgumentException("Not enough available seats for the match.");
            }
            for (int i = 0; i < diff; i++)
            {
                int seatNumber = match.getTotalSeats() - currentAvailable + i + 1;
                ticketRepository.save(new Ticket(client.getId(), match.getId(), "Seat " + seatNumber));
            }
        }
        else
        {
            removeTickets(client.getId(), match.getId(), Math.abs(diff));
        }
        match.setAvailableSeats(match.getAvailableSeats() - diff);
        matchRepository.update(match);
        updateObservers();
    }

    public List<ClientTicketDTO> getFilteredTickets(String name, String address)
    {
        return ticketRepository.getClientTicketsByCriteria(name, address);
    }
}
